source/tutorials/integration-testing-using-virtual-machines.md

```diff
@@ -2,43 +2,285 @@
+          firefox
+          thunderbird
+        ];
+      };
+
+      system.stateVersion = "22.11";
+    };
+    testScript = ''
+      machine.wait_for_unit("default.target")
+      machine.succeed("su -- alice -c 'which firefox'")
+      machine.fail("su -- root -c 'which firefox'")
+    '';
+  }
+```
+
+## Running tests
+
+To set up all machines and execute the test script:
+
+```shell-session
+$ nix-build minimal-test.nix
+```
+
+
+    ...
+    test script finished in 10.96s
+    cleaning up
+    killing machine (pid 10)
+    (0.00 seconds)
+    /nix/store/bx7z3imvxxpwkkza10vb23czhw7873w2-vm-test-run-minimal-test
+
 
+## Interactive Python shell to interact with virtual machine
+
+When developing tests or when something breaks, it’s useful to interactively tinker with the test or access a terminal for a machine.
+
+To interactively start a Python session with the testing framework:
+
+```shell-session
+$ $(nix-build -A driverInteractive minimal-test.nix)/bin/nixos-test-driver
+```
+
+You can run any of the testing operations.
+The `testScript` attribute from `minimal-test.nix` definition can be executed with `test_script()` function.
+
+Within this Python shell you can enter a interactive shell and run Python commands like those in the test script.
+
+If a virtual machine is not yet started, the test environment takes care of it on the first call of a method of a `machine` object.
+
+But you can also manually trigger the start of the virtual machine by using
+
+```shell-session
+>>> machine.start()
+```
+for a specific node,
+
+or
+
+```shell-session
+>>> start_all()
+```
+for all specified nodes.
+
+You can enter a interactive shell on the virtual machine using:
+
+```shell-session
+>>> machine.shell_interact()
+```
+
+and run commandline commands like:
+
+```shell-session
+uname -a
+```
+
+    Linux server 5.10.37 #1-NixOS SMP Fri May 14 07:50:46 UTC 2021 x86_64 GNU/Linux
+
+
+## Re-run successful tests
+
+Because test results are kept in the Nix store, a successful test is cached.
+This means that Nix will not run the test a second time as long as the test setup (node configuration and test script) stays semantically the same.
+Therefore, to run a test again, one needs to remove the result.
+
+If you would try to delete the result using the symbolic link, you will get the following error:
+
+```shell-session
+nix-store --delete ./result
+```
+
+
+    finding garbage collector roots...
+    0 store paths deleted, 0.00 MiB freed
+    error: Cannot delete path '/nix/store/4klj06bsilkqkn6h2sia8dcsi72wbcfl-vm-test-run-unnamed' since it is still alive. To find out why, use: nix-store --query --roots
+
+
+Instead, remove the symbolic link and only then remove the cached result:
+
+```shell-session
+rm ./result
+nix-store --delete /nix/store/4klj06bsilkqkn6h2sia8dcsi72wbcfl-vm-test-run-unnamed
+```
+
+This can be also done with one command:
+
+```shell-session
+result=$(readlink -f ./result) rm ./result && nix-store --delete $result
+```
```

Comment:
This section doesn't fit super well into this tutorial, it may be better fit for e.g. the Nix manual (`nix-build --check`, `nix-store --delete`), or some documentation on non-deterministic derivations in general.

@06kellyjac will add this to the Nix manual.